### PR TITLE
Keep Link UI open upon initial link creation when used in RichText

### LINKED ISSUE #53656: Inconsistent link preview flow between pasting and selecting links
When pasting a link onto text, we get the preview state right off. But when searching and clicking an internal entity to apply, LinkControl is closed entirely. Perhaps both should perform like pasting does, where the link preview is the next view you see. 

The effort in https://github.com/WordPress/gutenberg/pull/53566 attempts to do this at the moment. 

Linking via searching: 

https://github.com/WordPress/gutenberg/assets/1813435/221ec2de-e9f6-472a-b4ce-a769049085e1

Linking via paste: 

https://github.com/WordPress/gutenberg/assets/1813435/e1915c44-1b59-4e09-8728-dfd24618a871

### PR DESCRIPTION
**Pull request by getdave**
## What?

When a link is initially created within Rich Text, instead of automatically hiding the Link UI, this PR will keep it open to allow further editing of the newly created link (e.g. toggling link settings...etc).

Moreover, when toggling the `Opens in new tab` checkbox on the "Preview" state of the Link UI, the UI will no longer automatically close. 

Closes https://github.com/WordPress/gutenberg/issues/53656

## Why?

This allows users to create a link and then immediately edit it's properties without having to go through the process of re-activating the link they just created. This solves a number of issues and is better UX.

## How?

This PR has a number of changes which should be carefully assessed:

- `useAnchor` now exposes a `update` method on the returned `anchor` object. This is to allow consumers to force the anchor to be updated which may be required under very specific circumstances (such as in this PR). 
- Does not clear the activeFormats of the underlying `RichTextValue` when the Link UI is submitted, which means the link format remains active, which means that the Link UI **continues to be displayed**.
- Removes the `key` prop on the `LinkControl`. This was causing focus loss on initial link creation because the focus target was removed from the DOM when the component was remounted. This is because the `key` prop was computed because on the `anchor` returned from `useAnchor` which (due to the next UX requirements) now changes _whilst the `LInkControl` component is rendered_. I believe this forcable remount has not be necessary since we undertook improvements to the internals of `LinkControl` but I'm pinging @kevin940726 for a confidence check.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/444434/d3de6c3c-6d75-4502-96d2-b140714efed2

### CODE REVIEW

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [inline packages/rich-text/src/component/use-anchor.js]: I don't like providing dependencies to the effect in this way. It feels like an anti-pattern.

However, the bug it's solving is very real. Essentially if you have a `Popover` which is attached to element X and then you want that `Popover` to remain open and attached to element X when you focus _into_ the `Popover` the anchor is never re-computed. 

This is because when the focus moves away from element X into the popover, the "focus" event listeners get removed from element X as well. This means the anchor remains as it was when the popover first rendered.

**Comment by getdave** [comment]
Here is an example of what occurs if we don't force the re-computation of the popover anchor in `useAnchor`


https://github.com/WordPress/gutenberg/assets/444434/38c520ad-42fb-4196-b7d0-c28ef78c1752

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [inline packages/rich-text/src/component/use-anchor.js]: We probably can't really get away with modifying the API in this way. I expect we will have to create a complementary hook which basically does the same thing as `useAnchor` but has a different "export" signature. 

For example:

```
const { anchor, updateAnchor } = useMyNewAnchor( args );
```

Thoughts welcomed.

**Review by scruffian — DISMISSED** [review]
I think this is a good step in the right direction.

File: packages/rich-text/src/component/use-anchor.js
Comment (scruffian) [reply packages/rich-text/src/component/use-anchor.js]: I suppose we can't modify the return of `useAnchor`?

Comment (getdave) [reply packages/rich-text/src/component/use-anchor.js]: It's a public API so no.

I think we'll need a new hook.

We could pass additional dependencies into the hook to trigger the `useLayoutEffect` to recompute the value but that would couple the implementation to specific dependencies.

I think exposing the update function is the simplest IoC technique.

**Comment by getdave** [comment]
**Bugs** 

- when you initial create the link and then toggle "open in new tab" then the setting is not persisted. It _is_ however, on subsequent attempts once the link is already present.
- the first time you create the link there is a focus trap on the Popover. This was not intentional.
- click `Save` should not close the Popovoer and should go back to "Preview" state

**Review by jeryj — COMMENTED** [review]
Can we remove the tooltip from the Link Preview link? The full URL is already underneath it, and the tooltip covers it up:

Link Preview Link Tooltip covering url text

**Comment by richtabor** [comment]
> Can we remove the tooltip from the Link Preview link? The full URL is already underneath it, and the tooltip covers it up:

I say we do a follow-up soon to tidy up this preview in general (https://github.com/WordPress/gutenberg/issues/53654). Perhaps something like this: 

image

**Review by richtabor — DISMISSED** [review]
This feels much better. A step in the right direction for sure.

**Comment by getdave** [comment]
@jeryj ~Yeh I think we can ditch that.~ Actually I think this was added as a means to be able to see the entire URL if/when the URL is extremely long. We clip the URL for display reasons so folks needed a means to see the full URL. This was the compomise.

@richtabor yup I agree with refined previews.

I'll merge this after I fix [these bugs](https://github.com/WordPress/gutenberg/pull/57726#issuecomment-1887207984)

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [reply packages/rich-text/src/component/use-anchor.js]: I'm going to create a new base hook called `useAnchorBase` which will basically be `useAnchor` but it will export an object with the anchor and update method.

Then I'll rewrite `useAnchor` to call that hook but only return the `anchor` (as it does currently).

Then I"ll create a new private hook to also use the `useAnchorBase` but it will return the entire object.

Then I'll consume the private hook in this PR.

**Comment by richtabor** [comment]
This will also close https://github.com/WordPress/gutenberg/issues/53656

**Comment by liviopv** [comment]
I feel like this one will be well-received in general and I personally like it a lot, but it might also be seen as adding an extra click. @getdave if I type space or return after adding the link while the modal is open, does that action closes the modal and return the focus to the text cursor?

File: packages/private-apis/src/implementation.js
Comment (getdave) [inline packages/private-apis/src/implementation.js]: @ellatrix I'm opting RichText into privateAPIs. Just FYI

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [inline packages/rich-text/src/component/use-anchor.js]: We cannot modify the existing public API of the original `useAnchor` to expose an `update` method because it only returns a single object. That object can be interogated by consumers and thus it is not safe to expose additional properties on it (trust me I tried and it immediately broke the autocompleter tests).

The hook continues to behave exactly as it did before, except that the implementation has moved to the "base" hook and we simply call that and expose only the `anchor` from the `useAnchor` hook.

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [inline packages/rich-text/src/component/use-anchor.js]: This is our new private version of `useAnchor`. The only difference is that it exposes the full API of the "base" hook. The important part is that we expose the `update` method as part of the returned object.

We could selectively only expose `anchor` and `callback` here - i'm flexible about that.

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [reply packages/rich-text/src/component/use-anchor.js]: I've now exposed a private API. Let me know what you think @scruffian.

**Comment by getdave** [comment]
@joedolson @jeryj Testing the a11y I noticed that there appears to be a focus trap which is active on the initial creation of the link only.

When you first create the link your focus is trapped within the preview or the editor form and it is not possible to TAB out. Only ESC allows you to escape focus trap.

If you then re-activate the link the focus trap is no longer in evidence.

Here's a video demonstrating what I mean:

https://github.com/WordPress/gutenberg/assets/444434/fe8daa14-d2e2-4a21-9b25-ce1a8cf10ac1

I'm going to try and look into this but if you have any ideas what could be causing this it would be greatly appreciated.

Update: I suspect this line is the source of the bug:

https://github.com/WordPress/gutenberg/blob/7fce392ac694f4dc66ba6300f7b3356e9cdc2123/packages/compose/src/hooks/use-dialog/index.ts#L86

In `inline.js` we set conditionally set `focusOnMount`. This is to 

- allow the Link UI Popover to retrieve auto focus when you _trigger_ the Link UI for the initial creation of the Link.
- stop the `Popover` auto-focusing when you place your cursor into an existing link format. This is so you can use arrow keys to move into the link format without having focus "stolen" by the Link UI popover. 

However, by doing this we toggle the constrained tabbing behaviour in the `useDialog` hook. This is why we get different behaviour.

We'll need to work on this. For example, is it suitable to allow `Popover` to control whether it has a focus trap or not? That would also require modifying `useDialog` to allow for this override.

**Comment by getdave** [comment]
> when you initial create the link and then toggle "open in new tab" then the setting is not persisted. It is however, on subsequent attempts once the link is already present.

This bug also appears on `trunk` but does not manifest because the Link UI closes. If you tweak the code on `trunk` so the UI remains open then you can replicate this bug there also.

For some reason despite `onChange` being called with the correctly modified `value` (i.e. one where `openInNewTab` is set to `true`) when you re-activate the UI the `value` has reverted and lost that change. 

Update: the reason is in this line:

https://github.com/WordPress/gutenberg/blob/dbf74299c5245bdeee58cd9946db46ed060d027b/packages/format-library/src/link/inline.js#L133

The `formats` of the new `RichTextValue` are not set correctly because the start and end indices of `value` used in `applyFormat` are outside the boundary of the existing link format. This means the `startFormat` is never found here:

https://github.com/WordPress/gutenberg/blob/7ca256c2793a5d225381d29faea44832c58dbd53/packages/rich-text/src/apply-format.js#L39-L41

...which means the `formats` are never updated on the resulting `RichTextValue` which results in the new `opensInNewTab` not being persisted.

I think this is because on submit of the initial value the cursor is placed outside the new format which means it is no longer active. But in our case we want the cursor to remain within the format boundary because the format is still active. Here's a screen grab showing the cursor position after initial submit:

Screen Capture on 2024-01-12 at 11-36-40

**Comment by getdave** [comment]
I think looking at how the Text Color format works might provide clues on how to simplify the link format. For example, its UI remains open when you change different colors.

File: packages/rich-text/src/component/use-anchor.js
Comment (getdave) [reply packages/rich-text/src/component/use-anchor.js]: Actually I found an even better way which is standardised

https://github.com/WordPress/gutenberg/pull/57726/commits/160f4d3c46bae1f24b89dc4473e754b36c5c6622

File: packages/private-apis/src/implementation.js
Comment (getdave) [reply packages/private-apis/src/implementation.js]: I've reverted this as I discovered a new way which is already in use in the text color format:

https://github.com/WordPress/gutenberg/pull/57726/commits/160f4d3c46bae1f24b89dc4473e754b36c5c6622

File: packages/format-library/src/link/inline.js
Comment (getdave) [inline packages/format-library/src/link/inline.js:181]: This line was added by @ellatrix in https://github.com/WordPress/gutenberg/pull/17126.

I proposed we revert it. Why? The new UX that this PR is aiming to implement for links is to leave the Link UI open when first submitting a link. This provides an improvement because 

- users can validate they added the correct link
- users can modify the link settings after they added it

(further rationale is available on the linked Issue).

Unfortunately, by collapsing the selection and placing it _outside_ the link format we just created/edited we make it impossible for our APIs to determine how to apply further updates to the link format because it cannot "find" the format (because the format is no longer technically within the selection range).

This was what was causing [this bug](https://github.com/WordPress/gutenberg/pull/57726#issuecomment-1888860129) with `Opens in new tab` updating not being persisted.

When assessing whether or not we should drop this line, it's worth reviewing @WunderBart's [screenshots from his approval in 2020](https://github.com/WordPress/gutenberg/pull/17126#pullrequestreview-370485307) which show what UX this aims to achieve - namely that once a link is created the cursor moves outside the link to allow you to continue typing.

**Comment by getdave** [comment]
This PR is currently blocked by:

- ~[this UX issue and resulting bug](https://github.com/WordPress/gutenberg/pull/57726/files#r1450580242)~ I now consider this resolved as most folks seem to think the new UX is suitable.
- [this focus trap a11y issue and resulting bug](https://github.com/WordPress/gutenberg/pull/57726#issuecomment-1888800074)

File: packages/format-library/src/link/inline.js
Comment (draganescu) [reply packages/format-library/src/link/inline.js:181]: I think this change is good because linking is a format and all other formats keep the selection. The small speed gain that allows typing right after linking is not worth the trade off of not seeing confirmation of your applied format.

**Comment by joedolson** [comment]
So, I think that the trapped focus is potentially a good thing. With trapped focus, this control behaves like a modal dialog, just without a background mask. That may actually be beneficial to preventing keyboard users from getting lost. It becomes a problem because there's no cancel/close option unless you're in editing mode, which is non-intuitive. 

The question here is ultimately whether this control should have a modal behavior or not. I think that modal behavior is generally a good thing.

**Comment by jeryj** [comment]
> Testing the a11y I noticed that there appears to be a focus trap which is active on the initial creation of the link only.

I noticed this too, but I think it's the right behavior because of how you get to the dialog when creating a link (tab keypress vs invoked via button)

### Creating link
- Highlight text
- Press Link Button in toolbar/Command + K
- Opens link creation as dialog with focus trap 👍🏻 
- Create link
- Focus trapped in the preview. Good, if we want focus within the preview after creation, but I think it makes more sense to return focus to the place that created it. It's confusing behavior to have the focus still be within a modal.

A separate bug: **escape from the dialog should return you to the place that invoked the dialog.** Maybe there are instances this isn't a hard-and-fast rule though. @joedolson may have more info about this. When arriving at the Link Edit from the Toolbar, I think Escape should return focus back to the Toolbar.

### Link Preview Behavior
For the link preview, it shouldn't be a focus trap because you get to it via tabbing. If you tabbed into it and then it trapped focus, it would be an accessibility issue. You shouldn't arrive in a focus-trapping dialog that way. 

The issue with the link preview is that it isn't discoverable. How would a screen reader user know that they have access to the link preview via a Tab keypress when the text cursor is on the linked text? When I last tested, there was no announcement, and if there were an announcement, it would likely be quite annoying to have it announced. 

So, on one hand, it's an issue that it's not announced when it shows up, and also, it might be an annoyance if it's announced every time it shows up 😅 @alexstine will have better insight here on the balance of discoverability vs verbosity. 

### Editing a Link
For a screen reader user, IMO there isn't a discoverable way to edit a link. The toolbar includes an Unlink button, but no Edit Link button. To edit a link they'd need to know the Command + K convention (a common pattern, which is great) or know to press Tab when the carat is on the linked text.

To discern via if it should be a focus trap or not, I think it boils down to how you arrive at it (Tabbing vs Button/Command + K). In this situation code-wise in the current behavior, I think it can be simplified to: **Editing a link is a focus trap, Link Preview is not.**

The focus trap inconsistency is apparent when you Tab into the link preview:
- Tab to link preview 
- Not a focus trap 👍🏻 
- Edit link
- Currently not a focus trap. I think it should be a focus trap, and Cancel/Save returns you to the Link Preview

### Summary

- Link Creation: Focus goes back to where it came on cancel/create. Don't move focus into the preview after link creation.
- Link Editing: Should always be a focus trap
- Link Preview: Not a focus trap
- Link Preview: Not discoverable to screen readers in the current state.
- Exiting the focus trap should always return focus from whence you came (back to the toolbar, the text editor, or the link preview). The bug to test for here is Command + K on edited link goes to the link edit modal (great) and the Escape keypress puts you in the link preview (unexpected).
- Add/Change the Block Toolbar to have an Edit Link button in favor of/in addition to the Unlink button.

**Comment by getdave** [comment]
Also noting what Alex said: in https://github.com/WordPress/gutenberg/pull/54063#issuecomment-1708858002:

> When editing a link, you must use the shortcut. From what I can tell, there is only an unlink button in the formatting toolbar and there should also likely be an edit option there for best discovery. It makes little sense that by highlighting text, I would then tab to find a random edit button for my created link when tabbing under any other condition would land me in the sidebar.

File: packages/format-library/src/link/inline.js
Comment (scruffian) [reply packages/format-library/src/link/inline.js:181]: I think it makes sense to remove the change if we want to keep the Link UI open as the two things are connected - when we hide the UI we want to move the cursor to the end to enable you to carry on typing. However if we keep the UI open then we should leave the cursor where it is.

Comment (getdave) [reply packages/format-library/src/link/inline.js:181]: OK let's consider this resolved.

**Comment by getdave** [comment]
@jeryj we need to review https://github.com/WordPress/gutenberg/pull/57962

**Comment by getdave** [comment]
> The question here is ultimately whether this control should have a modal behavior or not. I think that modal behavior is generally a good thing.

I think that Modal behaviour _can_ be a good thing but possibly only in certain circumstances.

When a link has been created the "preview" of the link is purely informational. Currently you must tab into it and thus trapping focus within it would probably be a poor UX - you should be able to shift-tab to exit again.

However, once you enter "Edit" mode this is now a form. You should not be able to tab out of the form until you have submitted it.

So for me the goal here (albeit possibly a temporary one - see below) should be to [allow force constrained tabbing to be off in certain specific circumstances](https://github.com/WordPress/gutenberg/issues/57877).

It's worth noting that we have [an improved UX for perceiving and entering the control here](https://github.com/WordPress/gutenberg/issues/57821) which may make some of this discussion redundant. It tries to address the longstanding a11y concerns with accessing the control

**Review by draganescu — APPROVED** [review]
I think this PR solves a huge UX issue that was just too easy to overcome to be pressing but extremely annoying (closing the link ui immediately after creating the link).

I think Dave addressed all the original issues with the udpates to rich text and alos I don't think the focus trap issue (pre-existing) is blocking this PR.

File: packages/format-library/src/link/inline.js
Comment (getdave) [inline packages/format-library/src/link/inline.js]:
```suggestion
	// This ensures the Popover is positioned correctly on initial submission of the link.
```

**Comment by getdave** [comment]
Thanks for your review @draganescu 👍  

Noting that in order to merge this PR we'll need to address the e2e test failures and ensure they aren't failing due to valid regressions.

File: test/e2e/specs/editor/blocks/links.spec.js
Comment (getdave) [inline test/e2e/specs/editor/blocks/links.spec.js]: Should we invert this instead? `toBeVisible()`?

Comment (jeryj) [reply test/e2e/specs/editor/blocks/links.spec.js]: You'd need to do a little bit more, since the focus is in a different spot. I modified it to match the spirit of what the test was already testing. https://github.com/WordPress/gutenberg/pull/57726/commits/fb8f5a9d96075154ad61b13f5b14265b4574e26d

**Pull request merged**
